### PR TITLE
FileTarget - Faster async processing of LogEvents for the same file

### PR DESCRIPTION
**Pull request by snakefoot**
When performing async batch writing, then LogEvents  destination (file or database) can be decided by a Layout.

The default situation is that all LogEvents should go to the same destination. Have changed BucketSort() to optimize for this case.

This skips allocation of Dictionary and inserting all LogEvents into a new List of this dictionary

---

This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1683)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Cool!

Could you send a test program for this to https://github.com/NLog/NLog-Performance-tests ?

File: src/NLog/Internal/SortHelpers.cs [hunk 1]
```diff
@@ -61,14 +62,14 @@ internal static class SortHelpers
         /// <returns>
         /// Dictionary where keys are unique input keys, and values are lists of <see cref="AsyncLogEventInfo"/>.
         /// </returns>
-        public static Dictionary<TKey, List<TValue>> BucketSort<TValue, TKey>(this IEnumerable<TValue> inputs, KeySelector<TValue, TKey> keySelector)
+        public static IDictionary<TKey, IList<TValue>> BucketSort<TValue, TKey>(this IEnumerable<TValue> inputs, KeySelector<TValue, TKey> keySelector)
```
Comment (304NotModified) [inline src/NLog/Internal/SortHelpers.cs]: AFAIK the usage of interfaces is always slower than the direct type. Not sure if this is also for these cases.

Comment (304NotModified) [inline src/NLog/Internal/SortHelpers.cs]: Any idea about this?

**Comment by snakefoot** [comment]
When using the following setup with Tester1, then the new BucketSort gives 1.200.000 msg/sec (Before 1.000.000 msgs/sec)

>       <target xsi:type="BufferingWrapper"
>           name="mipLoggertarget"
>           slidingTimeout="false"
>           bufferSize="10000"
>           flushTimeout="0">
>         <target name ="mipLoggertargetWrapped" xsi:type="File"
>                 fileName="c:/temp/Log/test1.log"
>                 layout="${message}"
>                 keepFileOpen="true" AutoFlush="true" ConcurrentWrites="false" CleanupFileName="false" />
>       </target>

File: src/NLog/Internal/SortHelpers.cs (hunk 1 above)
Comment (snakefoot) [inline src/NLog/Internal/SortHelpers.cs]: Yes the IList prevents us from using the List-struct-enumerator. This is also why I have changed the iteration of a single bucket into a for-loop.

**Comment by 304NotModified** [comment]
That would be an very nice improvement!

**Comment by 304NotModified** [comment]
This also counts for the aysnc usage, isn't it?

**Comment by 304NotModified** [comment]
Reviewed 7 of 11 files at r1, 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by 304NotModified** [comment]
will try to test this, this week

**Comment by snakefoot** [comment]
Yes it will also improve the async-wrapper-performance. But because of the timer-throttle, then it is difficult to see this improvement. But together with #1669 (with sleeptime=0) then it should give a significant performance boost.

**Comment by 304NotModified** [comment]
Luckily  #1669  has been merged :)

**Comment by snakefoot** [comment]
Just merged with master to try the async-wrapper. It improved from 950.000 msgs/sec to 1.050.000 msgs/sec.

>       <target name="mipLoggertarget" xsi:type="AsyncWrapper" overflowAction="Block" queueLimit="40000" batchSize="20000" timeToSleepBetweenBatches="0">
>         <target name ="mipLoggertargetWrapped" xsi:type="File"
>                 fileName="c:/temp/Log/test1.log"
>                 layout="${message}"
>                 keepFileOpen="true" AutoFlush="true" ConcurrentWrites="false" CleanupFileName="false" />
>       </target>

**Comment by 304NotModified** [comment]
> Just merged with master to try the async-wrapper. This PR now improves from 950.000 msgs/sec to 1.050.000 msgs/sec.
> 
>   <target name="mipLoggertarget" xsi:type="AsyncWrapper" overflowAction="Block" queueLimit="40000" batchSize="20000" timeToSleepBetweenBatches="0">
>     <target name ="mipLoggertargetWrapped" xsi:type="File"
>             fileName="c:/temp/Log/test1.log"
>             layout="${message}"
>             keepFileOpen="true" AutoFlush="true" ConcurrentWrites="false" CleanupFileName="false" />
>   </target>
> When using timeToSleepBetweenBatches="1" then this PR improves performance from 800.000 msgs/sec. to 850.000 msgs/sec.

I have tested this case with master (639b40f803e3a3fbcd2de946a9d7a017181c7915) and pr/1683 (49d3e3d52e0f8abee57cda7a2d0479e87d4f6f60) and don't see any performance difference. Both will get around 240-260K/s. Any idea why?

> start test with 1.000.000 messages
> 1.000.000,00 messages. Time taken: 4.096,90ms. 244.086,89 / sec

Release build, https://github.com/NLog/NLog-Performance-tests/tree/7d3bbe83db1c42195c78cb313203b8f792e6019d

**Comment by snakefoot** [comment]
I have modified the Tester1 program to just do **logger.Debug("Line : 99999999");** , instead of adding extra allocations to the test.

I performed a test with the above async-wrapper config, and it improved from 1.000.000 msgs/sec to 1.100.000 msg/sec.

When using the buffer-wrapper-config then it improved from 1.000.000 msgs/sec to 1.200.000 msgs/sec.

If I add {threadname} into the layout (so the Logger-thread performs the rendering), then the buffer-wrapper improves from 900.000 msgs/sec to 1.000.000 msgs/sec.

**Comment by 304NotModified** [comment]
> I have modified the Tester1 program to just do logger.Debug("Line : 99999999"); , instead of adding extra allocations to the test.

So there is only a difference when using fixed values? 

Which is a valid case, as long as the non-fixed values (the one I have tested), won't suffer - (and that is not the case, no performance impact, also not lower)

**Comment by snakefoot** [comment]
There was also an improvement when adding {threadname} to the layout, so not just fixed values.

**Comment by snakefoot** [comment]
Removed code regarding ArraySegment, as this logic is for another PR

**Comment by 304NotModified** [comment]
Reviewed 11 of 11 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/NLog/Internal/SortHelpers.cs, line 95 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMXTzvwshHfGQJMpAF:-KVMXTzvwshHfGQJMpAG:boqvuff) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Internal/SortHelpers.cs#L95)):*
> ```Smalltalk
>         /// Dictionary where keys are unique input keys, and values are lists of <see cref="AsyncLogEventInfo"/>.
>         /// </returns>
>         public static ReadOnlySingleBucketDictionary<TKey, IList<TValue>> BucketSort<TValue, TKey>(this IList<TValue> inputs, KeySelector<TValue, TKey> keySelector)
> ```

Are you sure this code is used? I only see it in the tests? (reviewing in the webinterface)

---

*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMWryvD65ICm3WsC8A:b-ordl4w) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
> ```Smalltalk
>                     LogEventInfo firstLogEvent = null;
> 
>                     for (int i = 0; i < bucket.Value.Count; ++i)
> ```

why ++i and not i++?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by snakefoot** [comment]
*[src/NLog/Internal/SortHelpers.cs, line 95 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMXTzvwshHfGQJMpAF:-KVMYTM4IKfkTiKNYdAk:b-xm450i) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Internal/SortHelpers.cs#L95)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> Are you sure this code is used? I only see it in the tests? (reviewing in the webinterface)
</details>
It should have higher priority as it takes an IList (array) instead of IEnumerable. When debugging MultiFileWrite-test then it hits this method from FileTarget.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by snakefoot** [comment]
*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMYxg6DEMusDeIIwPj:b-xgkw2t) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> why ++i and not i++?
</details>
Old C++ habits are hard to kill :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by 304NotModified** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/NLog/Internal/SortHelpers.cs, line 95 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMXTzvwshHfGQJMpAF:-KVMZUjZulgeds91xKnA:bfmhfdb) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Internal/SortHelpers.cs#L95)):*
<details><summary><i>Previously, snakefoot (Rolf Kristensen) wrote…</i></summary>
> It should have higher priority as it takes an IList (array) instead of IEnumerable. When debugging MultiFileWrite-test then it hits this method from FileTarget.
</details>
A yes, that sounds logical

---

*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMZZ-__LYH3GSLFi5L:b-c96t70) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
<details><summary><i>Previously, snakefoot (Rolf Kristensen) wrote…</i></summary>
> Old C++ habits are hard to kill :)
</details>
LOL. Please change it :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by snakefoot** [comment]
*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMcN-6gxZwRFf9qQIm:b-v27ozw) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> LOL. Please change it :)
</details>
Guess you should make a file search for "++i") in the NLog-source (with your favorite index-variable-name). Apparently not the first :)

But will change it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by 304NotModified** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMd7ZJC4NTOpABUHo0:b-xqsmct) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
<details><summary><i>Previously, snakefoot (Rolf Kristensen) wrote…</i></summary>
> Guess you should make a file search for "++i") in the NLog-source (with your favorite index-variable-name). Apparently not the first :)
> 
> But will change it.
</details>
Will do if im bored :p.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by snakefoot** [comment]
*[src/NLog/Targets/FileTarget.cs, line 1019 at r5](https://reviewable.io:443/reviews/nlog/nlog/1683#-KVMWryvD65ICm3WsC89:-KVMdaiSIvL0o3PwAAxY:b-896fix) ([raw file](https://github.com/nlog/nlog/blob/8c4b28319c945605aeee1f5186d1d5413ef51668/src/NLog/Targets/FileTarget.cs#L1019)):*
<details><summary><i>Previously, 304NotModified (Julian Verdurmen) wrote…</i></summary>
> Will do if im bored :p.  
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Comment by 304NotModified** [comment]
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1683)*

**Pull request merged**